Userland/Applications/Calculator/Calculator.cpp

```diff
@@ -22,78 +26,117 @@ Crypto::BigFraction Calculator::begin_operation(Operation operation, Crypto::Big
     case Operation::Subtract:
     case Operation::Multiply:
     case Operation::Divide:
-        m_saved_argument = argument;
-        m_operation_in_progress = operation;
-        return argument;
+        m_binary_operation_saved_left_side = argument;
+        m_binary_operation_in_progress = operation;
+        m_current_value = argument;
+        break;
 
     case Operation::Sqrt:
         if (argument < Crypto::BigFraction {}) {
             m_has_error = true;
-            return argument;
+            m_current_value = argument;
+            break;
         }
-        res = argument.sqrt();
+        m_current_value = argument.sqrt();
         clear_operation();
         break;
     case Operation::Inverse:
         if (argument == Crypto::BigFraction {}) {
             m_has_error = true;
-            return argument;
+            m_current_value = argument;
+            break;
         }
-        res = argument.invert();
+        m_current_value = argument.invert();
         clear_operation();
         break;
     case Operation::Percent:
-        res = argument * Crypto::BigFraction { 1, 100 };
+        m_current_value = argument * Crypto::BigFraction { 1, 100 };
         break;
     case Operation::ToggleSign:
-        res = -argument;
+        m_current_value = -argument;
         break;
 
     case Operation::MemClear:
         m_mem.set_to_0();
-        res = argument;
+        m_current_value = argument;
         break;
     case Operation::MemRecall:
-        res = m_mem;
+        m_current_value = m_mem;
         break;
     case Operation::MemSave:
         m_mem = argument;
-        res = argument;
+        m_current_value = argument;
         break;
     case Operation::MemAdd:
         m_mem = m_mem + argument; // avoids the need for operator+=()
-        res = m_mem;
+        m_current_value = m_mem;
+        break;
+    case Operation::Equals:
+        m_current_value = argument;
         break;
     }
 
-    return res;
+    return m_current_value;
+}
+
+static bool operation_is_binary(Calculator::Operation operation)
+{
+    switch (operation) {
+
+    case Calculator::Operation::Add:
+    case Calculator::Operation::Subtract:
+    case Calculator::Operation::Multiply:
+    case Calculator::Operation::Divide:
+        return true;
+
+    default:
+        return false;
+    }
 }
 
-Crypto::BigFraction Calculator::finish_operation(Crypto::BigFraction argument)
+Optional<Crypto::BigFraction> Calculator::operation_without_argument(Operation operation)
+{
+    bool in_binary_operation = m_binary_operation_in_progress != Operation::None;
+    bool entering_new_binary_operation = operation_is_binary(operation);
+    bool previous_operation_was_binary = operation_is_binary(m_previous_operation);
+
+    if (in_binary_operation && entering_new_binary_operation) {
```

Comment:
Small nitpick on the style.

1. We usually don't put brackets around single line `if` and `else`
2. We don't use `else` after a `return`

Reply:
Done